src/cargo/ops/registry/auth.rs

```diff
@@ -0,0 +1,229 @@
+        run_command(config, process, registry_name, api_url, Action::Erase)?;
+    } else {
+        config::save_credentials(config, None, registry_name)?;
+    }
+    Ok(())
+}
+
+fn run_command(
+    config: &Config,
+    process: &(PathBuf, Vec<String>),
+    name: &str,
+    api_url: &str,
+    action: Action,
+) -> CargoResult<Option<String>> {
+    let cred_proc;
+    let (exe, args) = if process.0.to_str().unwrap_or("").starts_with("cargo:") {
+        cred_proc = sysroot_credential(config, process)?;
+        &cred_proc
+    } else {
+        process
+    };
+    if !args.iter().any(|arg| arg.contains("{action}")) {
+        let msg = |which| {
+            format!(
+                "credential process `{}` cannot be used to {}, \
+                 the credential-process configuration value must pass the \
+                 `{{action}}` argument in the config to support this command",
+                exe.display(),
+                which
+            )
+        };
+        match action {
+            Action::Get => {}
+            Action::Store(_) => bail!(msg("log in")),
+            Action::Erase => bail!(msg("log out")),
+        }
+    }
+    let action_str = match action {
+        Action::Get => "get",
+        Action::Store(_) => "store",
+        Action::Erase => "erase",
+    };
+    let args: Vec<_> = args
+        .iter()
+        .map(|arg| {
+            arg.replace("{action}", action_str)
+                .replace("{name}", name)
+                .replace("{api_url}", api_url)
+        })
+        .collect();
+
+    let mut cmd = Command::new(&exe);
+    cmd.args(args)
+        .env("CARGO", config.cargo_exe()?)
+        .env("CARGO_REGISTRY_NAME", name)
+        .env("CARGO_REGISTRY_API_URL", api_url);
+    match action {
+        Action::Get => {
+            cmd.stdout(Stdio::piped());
+        }
+        Action::Store(_) => {
+            cmd.stdin(Stdio::piped());
+        }
+        Action::Erase => {}
+    }
+    let mut child = cmd.spawn().chain_err(|| {
+        let verb = match action {
+            Action::Get => "fetch",
+            Action::Store(_) => "store",
+            Action::Erase => "erase",
+        };
+        format!(
+            "failed to execute `{}` to {} authentication token for registry `{}`",
+            exe.display(),
+            verb,
+            name
+        )
+    })?;
+    let mut token = None;
+    match &action {
+        Action::Get => {
+            let mut buffer = String::new();
+            log::debug!("reading into buffer");
+            child
+                .stdout
+                .as_mut()
+                .unwrap()
+                .read_to_string(&mut buffer)
+                .chain_err(|| {
+                    format!(
+                        "failed to read token from registry credential process `{}`",
+                        exe.display()
+                    )
+                })?;
+            if let Some(end) = buffer.find('\n') {
```

Comment:
I'd expect a call to `.trim()` here, but is this intentionally ignoring the second-and-after lines of output?

Reply:
Hm, yea, I changed it so that it will return an error if there is more than one line.